### PR TITLE
Add additional Windows-specific APIs

### PR DESCRIPTION
**Pull request by adamsitnik**
A follow up of https://github.com/dotnet/designs/pull/142

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 1]
```diff
@@ -220,6 +222,7 @@ public sealed partial class ComEventInterfaceAttribute : System.Attribute
         public System.Type EventProvider { get { throw null; } }
         public System.Type SourceInterface { get { throw null; } }
     }
+    [System.Runtime.Versioning.SupportedOSPlatformAttribute("windows7.0")]
```
Comment (jkotas) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs]: Is it intentional that some of the attributes use `windows7.0` and some just `windows` ?

Comment (adamsitnik) [reply src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs]: @jkotas We should not use OS version for marking stuff as Windows-specific, this is why I've found all such usages and fixed them in https://github.com/dotnet/runtime/pull/40610/commits/5b1e62c7a5f87976f045455804e11e6e92760f24

**Comment by terrajobst** [comment]
I pwrote a tool](https://github.com/terrajobst/HarvestPlatformSupport#differences-from-spec) to extract these attributes from reference assemblies. I've looked at the report and the only thing that is missing from spec is this: `System.Security.Cryptography.CspParameters`. Was that intentional?

/cc @bartonjs

**Review by terrajobst — APPROVED** [review]

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs [hunk 2]
```diff
@@ -139,6 +141,7 @@ private struct ComInterfaceInstance
         /// this <see cref="ComWrappers" /> instance, the previously created COM interface will be returned.
         /// If not, a new one will be created.
         /// </remarks>
+        [SupportedOSPlatform("windows")]
```
Comment (eerhardt) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:144]: Can we just mark the whole type as windows specific?

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 3]
```diff
@@ -401,6 +405,7 @@ public sealed partial class FileSystem
         public static void MkDir(string Path) { }
         public static void Print(int FileNumber, params object[] Output) { }
         public static void PrintLine(int FileNumber, params object[] Output) { }
+        [System.Runtime.Versioning.SupportedOSPlatformAttribute("windows")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:408]: Rename is windows specific? That seems odd.

Comment (eerhardt) [reply src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:408]: I looked at the implementation, it does throw PNSE on non-windows. Weird.

Comment (terrajobst) [reply src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:408]: Welcome to `My` 😁

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs (hunk 2 above)
Comment (terrajobst) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:144]: Seems reasonable to me

Comment (adamsitnik) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs]: done. 

BTW how should we handle nested types? If `class B` is nested within `class A` which is Windows-specific, should `B` be Windows-specific OOTB as well?

```cs
[SupportedOSPlatform("windows")]
public class A
{
    public class B { }
}
```

/cc @buyaa-n

Comment (eerhardt) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs]: >  If class B is nested within class A which is Windows-specific, should B be Windows-specific OOTB as well?

Yes, I would assume so. The same with any other scope - if a class is marked OS-specific, all the methods in it are OS-specific. If an assembly is marked OS-specific, all the types in it are OS-specific.

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/ComAwareEventInfo.cs [hunk 4]
```diff
@@ -19,6 +20,7 @@ public ComAwareEventInfo(Type type, string eventName)
             _innerEventInfo = type.GetEvent(eventName)!;
         }
 
+        [SupportedOSPlatform("windows")]
```
Comment (eerhardt) [inline src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/ComAwareEventInfo.cs:23]: It seems like the whole `ComAwareEventInfo` class should be marked as Windows-specific. In general, nothing "COM" specific will work outside of Windows.

**Pull request merged**